api/python/cellxgene_census/src/cellxgene_census/experimental/_embedding_search.py

```diff
@@ -0,0 +1,167 @@
+    assert index_metadata[0]["type"] == "IVFFlat", "Only IVFFlat index is supported (update cellxgene_census)"
+    mirrors = _get_census_mirrors()
+    mirror = mirror or cast(str, mirrors["default"])
+    mirror_info = cast(CensusMirror, mirrors[mirror])
+    uri = _uri_join(mirror_info["embeddings_base_uri"], index_metadata[0]["relative_uri"])
+    return uri, cast(str, mirror_info["region"])
+
+
+def predict_obs_metadata(
+    organism: str,
+    census_version: str,
+    neighbors: NeighborObs,
+    column_names: Sequence[str],
+    experiment: Optional[soma.Experiment] = None,
+) -> pd.DataFrame:
+    """Predict obs metadata attributes for the query cells based on the embedding nearest neighbors.
+
+    Args:
+        organism, census_version:
+            Embedding information as supplied to :func:`find_nearest_obs`.
+        neighbors:
+            Results of a :func:`find_nearest_obs` search.
+        column_names:
+            Desired obs metadata column names. The current implementation is suitable for
+            categorical attributes (e.g. cell_type, tissue_general).
+        experiment:
+            Open handle for the relevant SOMAExperiment, if available (otherwise, will be opened
+            internally). e.g. ``census["census_data"]["homo_sapiens"]`` with the relevant Census
+            version.
+
+    Returns:
+        Pandas DataFrame with the desired column predictions. Additionally, for each predicted
+        column ``col``, an additional column ``col_confidence`` with a confidence score between 0
+        and 1.
+    """
+    with ExitStack() as cleanup:
+        if experiment is None:
+            # open Census transiently
+            census = cleanup.enter_context(open_soma(census_version=census_version))
+            experiment = census["census_data"][_get_experiment_name(organism)]
+
+        # fetch the desired obs metadata for all of the found neighbors
+        neighbor_obs = (
+            experiment.obs.read(
+                coords=(neighbors.neighbor_ids.flatten(),), column_names=(["soma_joinid"] + list(column_names))
+            )
+            .concat()
+            .to_pandas()
+        ).set_index("soma_joinid")
+
+        # step through query cells to generate prediction for each column as the plurality value
+        # found among its neighbors, with a confidence score based on the simple fraction (for now)
+        # TODO: something more intelligent for numeric columns! also use distances, etc.
+        out: Dict[str, List[Any]] = {}
+        for i in range(neighbors.neighbor_ids.shape[0]):
+            neighbors_i = neighbor_obs.loc[neighbors.neighbor_ids[i]]
+            for col in column_names:
+                col_value_counts = neighbors_i[col].value_counts(normalize=True)
+                out.setdefault(col, []).append(col_value_counts.idxmax())
+                out.setdefault(col + "_confidence", []).append(col_value_counts.max())
+
+    return pd.DataFrame.from_dict(out)
```

Comment:
```suggestion
        max_joinid = neighbor_obs.index.max()
        out: dict[str, pd.Series] = {}
        indices = np.broadcast_to(np.arange(neighbors.neighbor_ids.shape[0]), (10, neighbors.neighbor_ids.shape[0])).T
        g = sparse.csr_matrix(
            (
                np.broadcast_to(1, neighbors.neighbor_ids.shape[0] * 10),
                (
                    indices.flatten(),
                    neighbors.neighbor_ids.astype(np.int64).flatten(),
                ),
            ),
            shape=(neighbors.neighbor_ids.shape[0], max_joinid + 1)
        )
        for col in column_names:
            col_categorical = neighbor_obs[col].astype("category")
            joinid2category = sparse.coo_matrix(
                (
                    np.broadcast_to(1, len(neighbor_obs)),
                    (
                        neighbor_obs.index,
                        col_categorical.cat.codes
                    )
                ),
                shape=(max_joinid + 1, len(col_categorical.cat.categories))
            )
            counts = g @ joinid2category
            rel_counts = counts / counts.sum(axis=1)
            out[col] = col_categorical.cat.categories[rel_counts.argmax(axis=1).A.flatten()].astype(object)
            out[f"{col}_confidence"] = rel_counts.max(axis=1).A.flatten()

    return pd.DataFrame.from_dict(out)
```

This should be faster, but accomplish the same thing. Unfortunately it looks like the results are a little different due to a difference in how ties are resolved (but that's only for low confidence labels anyways).

For 2.3k cells, using the current implementation:

```python
%%time
result = cellxgene_census.experimental.predict_obs_metadata(organism=organism, census_version=census_version, neighbors=test_query, column_names=["cell_type"])
```

```
CPU times: user 16.2 s, sys: 2.06 s, total: 18.3 s
Wall time: 15.2 s
```

For the implementation suggested here:

```python
%%time
result_new = predict_obs_metadata(organism=organism, census_version=census_version, neighbors=test_query, column_names=["cell_type"])
```


```
CPU times: user 3.21 s, sys: 1.86 s, total: 5.07 s
Wall time: 1.95 s
```